#Nicole E Soltis 
#from script by Jason A Corwin, Modified by Rachel Fordyce
#to run bigRR on Linux GPU for GWAS
#---------------------------------------------------------------

rm(list=ls())
setwd("~/Documents/GitRepos/BcSolGWAS/")
setwd("~/Projects/BcSolGWAS/")
############################################################################
###Plotting the HEM results

#NEED TO CHECK CONTIGS FOR THIS

#Load plotting package
library(ggplot2)
library(grid)

#Import data (reorganized from script ReformatBigRRouts.R)
HEM.plotdata <- read.csv("data/GWAS_files/04_bigRRoutput/SNP_overlap/TopSNPs_alltraits_golong.csv")

HEM.plotdata$Pos <- as.character(HEM.plotdata$Pos)#ensure that position data is not in scientific notation

#Reformat Chromosomes and Positions
HEM.plotdata$Chrom <- gsub("Chromosome", "", HEM.plotdata$Chrom)
HEM.plotdata$Chrom <- as.numeric(as.character(HEM.plotdata$Chrom))
HEM.plotdata$Segment <- as.numeric(as.character(HEM.plotdata$Segment))
HEM.plotdata$Pos <- as.numeric(as.character(HEM.plotdata$Pos))

#sort dataframe rows in order of Chrom, then Pos
HEM.plotdata2 <- HEM.plotdata[with(HEM.plotdata, order(Chrom, Segment, Pos)), ]

#Make plotting variables
HEM.plotdata$Index = NA
ticks = NULL
lastbase = 0

#want to figure out where to add +500 to draw breaks between chromosomes
#Redo the positions to make them sequential		-- accurate position indexing
#still need to fix plotting so that CONTIGS are sequential (HEM.plotdata$Segment)
for (i in unique(HEM.plotdata$Chrom)) {
  print(i)
  if (i==1) {
    HEM.plotdata[HEM.plotdata$Chrom==i, ]$Index=HEM.plotdata[HEM.plotdata$Chrom==i, ]$Pos
  }	else {
    #changed lastbase+tail to lastbase+max
    lastbase=+lastbase+max(subset(HEM.plotdata,HEM.plotdata$Chrom==i-1)$Pos, 1)
    HEM.plotdata[HEM.plotdata$Chrom==i, ]$Index=HEM.plotdata[HEM.plotdata$Chrom==i, ]$Pos+lastbase
  }
  ticks=c(ticks, HEM.plotdata[HEM.plotdata$Chrom==i, ]$Index[floor(length(HEM.plotdata[HEM.plotdata$Chrom==i, ]$Index)/2)+1])
}
ticklim=c(min(HEM.plotdata$Index),max(HEM.plotdata$Index))

write.csv(HEM.plotdata, "data/GWAS_files/04_bigRRoutput/SNP_overlap/TopSNPs_alltraits_Indexed.csv")

#now go long to wide
library(reshape2)
names(HEM.plotdata)
plotdata.wide <- reshape(HEM.plotdata, idvar=c("Chrom", "Segment", "Pos", "Index"), timevar="Trait", direction = "wide")
names(plotdata.wide)
plotdata.wide$Domest.Num <- rowSums(!is.na(plotdata.wide[,c(6,7,12)]))
plotdata.wide$Plant.Num <- rowSums(!is.na(plotdata.wide[,c(5,8,9,10,11,13,14,15,16,17,18)]))
plotdata.wide$Trait.Num <- rowSums(!is.na(plotdata.wide[,5:18]))
plotdata.wide$Phenos <- "NA"
plotdata.wide$Phenos <- ifelse(plotdata.wide$Plant.Num >= 1 & plotdata.wide$Domest.Num >= 1, "Both", ifelse(plotdata.wide$Plant.Num >= 1, "IndPlants", "Domestication"))

write.csv(plotdata.wide, "data/GWAS_files/04_bigRRoutput/SNP_overlap/TopSNPs_alltraits_wide.csv")

#make plots for each phenotype

#create a custom color scale for black and white version
#myColors <- c("grey40", "grey60", "grey40", "grey60", "grey40", "grey60", "grey40", "grey60", "grey40", "grey60", "grey40", "grey60", "grey40", "grey60", "grey40", "grey60")
#names(myColors) <- levels(HEM.plotdata$Chrom)
#colScale <- scale_colour_manual(name = "Chrom",values = myColors)

names(plotdata.wide)
#without the loop
jpeg("plots/MultiPlot/meta/Sl_LesionSize_MAF20_meta.ManhattanPlot.jpg", width=8, height=4, units='in', res=600)
  ggplot(plotdata.wide, aes(x=Index, y=abs(Trait.Num)))+
    theme_bw()+
#    colScale+ #add this for BW version only
    geom_point(aes(color = factor(Phenos)))+
    labs(list(y="Number of Phenotypes", x="Chromosome position", title="Number of Phenotypes with Sig Fx"))+
    guides(col = guide_legend(nrow = 8, title="Traits"))+
#    geom_hline(yintercept=get(paste("TH95_", names(HEM.plotdata[14]), sep="")), colour = "black") +
#    geom_text(aes(0,get(paste("TH95_", names(HEM.plotdata[14]), sep="")), label = ".95 Threshold", vjust = 1.5, hjust = .05), col = "black")+
#    geom_hline(yintercept=get(paste("TH99_", names(HEM.plotdata[14]), sep=""))) +
#    geom_text(aes(0,get(paste("TH99_", names(HEM.plotdata[14]), sep="")), label = ".99 Threshold", vjust = 1.5, hjust = .05), col = "black")+
   expand_limits(y=0)
dev.off()

jpeg(paste("plots/MultiPlot/NewModel0711b/Sl_LesionSize_MAF20_highTR_", names(HEM.plotdata[13]), ".ManhattanPlot.jpg", sep=""), width=8, height=4, units='in', res=600)
ggplot(HEM.plotdata, aes(x=Index, y=abs(HEM.plotdata[13])))+
  theme_bw()+
  colScale+
  geom_point(aes(color = factor(Chrom)))+
  labs(list(y="SNP Effect Estimate", x="Chromosome position", title=paste("Lesion Size on ", names(HEM.plotdata[13]))))+
  guides(col = guide_legend(nrow = 8, title="Chromosome"))+
  geom_hline(yintercept=get(paste("TH999_", names(HEM.plotdata[13]), sep=""))) +
  geom_text(aes(0,get(paste("TH999_", names(HEM.plotdata[13]), sep="")), label =
  ".999 Threshold", vjust = 1.5, hjust = .05), col = "black")+
  expand_limits(y=-0.001)
dev.off()
}
#stop [3]

#or just do them each by hand
jpeg("plots/Sl_LesionSize_greyIntx.jpeg", width=6, height=4, units='in', res=600)
ggplot(HEM.plotdata, aes(x=Index, y=abs(LA4345)))+
  labs(list(y="SNP Effect Estimate", title = "Lesion Size on LA4345", x="Chromosome position"))+
  theme_bw()+
  geom_point(aes(color = factor(Chrom)))+
  guides(col = guide_legend(nrow = 8, title="Chromosome"))+
  geom_hline(yintercept=TH95_LA4345, colour = "blue") +
  geom_text(aes(0,TH95_LA4345, label = ".95 Threshold", vjust = 1.5, hjust = .05), col = "blue")+
  geom_hline(yintercept=TH99_LA4345) +
  geom_text(aes(0,TH99_LA4345, label = ".99 Threshold", vjust = 1.5, hjust = .05), col = "black")
dev.off()

#geom_hline(yintercept=TH999_LA3475) +
#  geom_text(aes(0,TH999_LA3475, label = ".999 Threshold", vjust = 1.5, hjust = .05), col = "black")




jpeg("plots/Sl_LesionSize_LA0410.ManhattanPlot_999.jpg")
qplot(Index,abs(LA0410), data=HEM.plotdata, ylab="SNP Effect Estimate" , 
      main = "LesionSize_LA0410", colour=factor(Chrom)) +
  geom_hline(yintercept=TH999_LA0410) +
  geom_text(aes(0,TH999_LA0410, label = ".999 Threshold", vjust = 1.5, hjust = .05), col = "black")
dev.off()

#how many SNPs are above a certain threshhold?
topSNP <- sum(HEM.plotdata$LA0410 >= 0.001) #41
highSNP <- sum(HEM.plotdata$LA0410 >= TH999_LA0410) #spits out number
totalSNP <- sum(HEM.plotdata$LA0410 >= 0)
highSNP/totalSNP*100

#rest of plots
jpeg("plots/Sl_LesionSize_LA1547.ManhattanPlot_95.jpg")
qplot(Index,abs(LA1547), data=HEM.plotdata, ylab="SNP Effect Estimate" , 
      main = "LesionSize_LA1547", colour=factor(Chrom)) +
geom_hline(yintercept=TH95_LA1547, colour = "blue") +
  geom_text(aes(0,TH95_LA1547, label = ".95 Threshold", vjust = 1.5, hjust = .05), col = "blue")
dev.off()

###############################
###Make a BiPlot for two phenotypes (LA0410 and LA0480)
#This code has several dependent objects from the single HEM plotting code above

#set my colors
mycols=rep(c("gray10","gray60"),max(HEM.plotdata$Chrom))

for(k in c(3:7)) {
  #Make temp data for figure
  print(names(thresh.HEM[[1]])[k-2])
  
  CamLes.plot.data<-cbind("Cam",HEM.plotdata[,k],HEM.plotdata[,c(1,13)])
  colnames(CamLes.plot.data)[1:2] <- c("Pheno","Effect")
  CamLes.plot.data <- rbind(CamLes.plot.data,cbind(Pheno="Lesion",Effect=-HEM.plotdata[,k+5],HEM.plotdata[,c(1,13)]))
  CamLes.plot.data <- cbind(CamLes.plot.data,PhenoChrom=paste(CamLes.plot.data[,1],CamLes.plot.data[,3]))
  
  #Scale the effects and threshold
  temp.scaled <- scale(c(thresh.HEM[[3]][k-2],CamLes.plot.data$Effect[CamLes.plot.data$Pheno == "Cam"]))
  Cam.Scaled.Thresh <- abs(temp.scaled[1])
  CamLes.plot.data$Effect[CamLes.plot.data$Pheno == "Cam"] <- abs(temp.scaled[-1])
  
  temp.scaled <- scale(c(thresh.HEM[[3]][k+3],CamLes.plot.data$Effect[CamLes.plot.data$Pheno == "Lesion"]))
  Les.Scaled.Thresh <- -abs(temp.scaled[1])
  CamLes.plot.data$Effect[CamLes.plot.data$Pheno == "Lesion"] <- -abs(temp.scaled[-1])
  
  rm(temp.scaled)
  
  
  #Plot Data
  
  mycols2=c("royalblue","royalblue4","royalblue","royalblue4","royalblue","olivedrab3","olivedrab","olivedrab3","olivedrab","olivedrab3")
  
  
  if(grepl("Ctrl",colnames(HEM.plotdata)[k])){
    plot2=qplot(pos,Effect,data=CamLes.plot.data[CamLes.plot.data$Pheno=="Cam",], ylab="|Scaled SNP Effect Estimate|" , colour=factor(PhenoChrom))
    
    ##All the thresholds
    #thresholds <- c(thresh.HEM[[1]][k-2],thresh.HEM[[2]][k-2],thresh.HEM[[3]][k-2],thresh.HEM[[4]][k-2])
    #thresholds <- c(thresh.HEM[[3]][k-2])
    
    plot2=plot2 + geom_hline(yintercept=Cam.Scaled.Thresh, colour="red")
    
  } else {
    plot2=qplot(pos,Effect,data=CamLes.plot.data, ylab="|Scaled SNP Effect Estimate|" , colour=factor(PhenoChrom))
    ##All the thresholds
    #thresholds <- c(thresh.HEM[[1]][c(k-2,k+3)],thresh.HEM[[2]][c(k-2,k+3)],thresh.HEM[[3]][c(k-2,k+3)],thresh.HEM[[4]][c(k-2,k+3)])
    #thresholds <- c(thresh.HEM[[3]][c(k-2,k+3)])
    #thresholds[grepl("Lesion",names(thresholds))] <- thresholds[grepl("Lesion",names(thresholds))] * -1
    plot2=plot2 + geom_hline(yintercept=c(Cam.Scaled.Thresh,Les.Scaled.Thresh), colour="red")
    
    
  }
  #plot2=qplot(pos,Effect,data=CamLes.plot.data, ylab="SNP Effect Estimate" , colour=factor(PhenoChrom))
  plot2=plot2+scale_x_continuous(name="Chromosome", breaks=ticks, labels=(unique(HEM.plotdata$Chrom)))
  #plot2=plot2+scale_y_continuous(limits=c(0,maxy), breaks=1:maxy, labels=1:maxy)
  plot2=plot2+scale_colour_manual(values=mycols2) + theme(legend.position="none", axis.text.x=element_text(colour="black"), axis.text.y=element_text(colour="black"))
  #plot2=plot2 + geom_hline(yintercept=c(as.numeric(thresh.HEM[k-2]),-as.numeric(thresh.HEM[k+3])), colour="red")
  
  #write to file.  May want something in a vector graphic format
  jpeg(filename = paste(strsplit(names(thresh.HEM[[1]][k-2]),"_")[[1]][2],".jpeg",sep=""), width = 1500, height = 800)
  print(plot2)
  dev.off()
}



###############################
###Make a HEM plot of a region (Cam and Lesion)
#This code has several dependent objects from the single HEM plotting code above

#PRLM3 (Chr4 Pos 9560135:9565454) in this example.  Just change the Chrom and positions to find another gene.
Reg.min <- 9560135 - 50000
Reg.max <- Reg.min + 1000000


Reg.min <- 13750000 - 50000
Reg.max <- Reg.min + 100000


Region2Plot <- HEM.plotdata[HEM.plotdata$Chrom == 4 & HEM.plotdata$Pos %in% c(Reg.min:Reg.max),]
Region2Plot <- Region2Plot[,-match("pos",colnames(Region2Plot))]
Region2Plot <- melt(Region2Plot, id=c("Chrom","Pos"), variable.name = "Isolate", value.name = "Effect")
Region2Plot$Pheno <- NA
Region2Plot$Pheno[grepl("Cam",Region2Plot$Isolate)] <- "Camalexin"
Region2Plot$Pheno[grepl("Lesion",Region2Plot$Isolate)] <- "Lesion"


Region2Plot$Isolate <- gsub("Camalexin.ng.LesPer._|Lesion.Size.mm2_|.HEM","",Region2Plot$Isolate)

plot.reg <- qplot(Pos,abs(as.numeric(Effect)),data=Region2Plot[Region2Plot$Pheno == "Camalexin",], ylab="|Scaled SNP Effect Estimate|" , colour=factor(Isolate))

#Missing a gene at 13755649-13759740; At4g27550 (wouldn't print)
gene.start = c(13759841,13763464,13765661,13767763,13771185,13772819)
gene.stop = c(13761559,13765064,13766597,13769961,13772594,13777290)

plot.reg <- plot.reg + geom_segment(aes(x = gene.start, y = -0.001, xend = gene.stop, yend = -.001), colour = "red")

#####
###For some reason, I had to use the following code at the terminal to get this to print right.
#xwd > PolyMorphicRegOnChr4.xwd
#convert PolyMorphicRegOnChr4.xwd PolyMorphicRegOnChr4.jpg

jpeg(filename = "PolyMorphicRegionOnChr4.jpg", width = 1500, height = 800)
print(plot2)
dev.off()